### ISSUE TITLE
Online inference script of LAM model

### ISSUE THREAD

**Issue by lucasjinreal** [OpenDriveLab/UniVLA#13]
Hi, I'm wondering:

1) is there a script to infer the LAM model only with real-world videos? I want to see how it behaves for a simple task like navigation;
2) How to use the existing LAM to easily train a Qwen2.5 VL based two - system model? Currently it heavily relies on primisma which is a little complicated compared with SOTA vlms.

**Comment by retsuh-bqw** [comment]
LAM does not need to resolve or interpolate them. LAM is only to generate pseudo labels for VLM pretraining. I believe your are referring to the action deocder, which is built upon a multi-head attention pooling module and a MLP to directly decode latent action embeddings (aka, the VLM's last-layer faeture corresponding to latent action tokens, not the decoded <ACT_i> tokens in the form of text) into low-level action chunks with the L1 regression loss. So LAM is NOT used to decode the latent actions and it is completly deprecated during inference time.

**Comment by lucasjinreal** [comment]
hi, does the navigation task is accurate and robust? what if target is missing a long time, will it resume instructions?

**Comment by Trav1slaflame** [comment]
> ### 1. You can try the following script:
> import torchvision
> from latent_action_model.genie.modules.lam import ControllableDINOLatentActionModel
> 
> # Load model and checkpoint
> latent_action_model = ControllableDINOLatentActionModel(
>     in_dim=3,
>     model_dim=cfg.lam_model_dim,
>     latent_dim=cfg.lam_latent_dim,
>     num_latents=cfg.codebook_size,
>     patch_size=cfg.lam_patch_size,
>     enc_blocks=cfg.lam_enc_blocks,
>     dec_blocks=cfg.lam_dec_blocks,
>     num_heads=cfg.lam_num_heads,
>     dropout=0.,
> )
> 
> # Process and load checkpoint
> lam_ckpt = torch.load(cfg.lam_path)['state_dict']
> latent_action_model.load_state_dict({k.replace("lam.", ""): v for k, v in lam_ckpt.items()}, strict=True)
> latent_action_model = latent_action_model.to(device_id).eval()
> 
> # Process images and get latent action
> image_transform_lam = torchvision.transforms.ToTensor()  # DINO normalization handled by model
> video = torch.stack([
>     image_transform_lam(img),    # current observation
>     image_transform_lam(img_k)   # 1-sec after observation
> ], dim=0).unsqueeze(0).to(latent_action_model.device)
> 
> action_idx = latent_action_model.vq_encode(video)['indices'].squeeze()
> ### 2. This approach can be integrated with any VLM. In UniVLA, we simply add specialized tokens to the LLaMA tokenizer:
> special_tokens_dict = {'additional_special_tokens': [f'<ACT_{i}>' for i in range(cfg.codebook_size)]}
> num_added_toks = action_tokenizer.add_special_tokens(special_tokens_dict)
> During policy training for next-token prediction, map the latent action indices to their corresponding tokens:
> 
> action_vocab = [f'<ACT_{i.item()}>' for i in action_idx]
> The LLM tokenizer can then process these mapped latent action tokens normally.

Dear author,

Thanks for your wonderful work! I just want to quickly check is there a resolution limit for RGB images input into the LAM? Or it can deal with any regular image resolution (e.g., 200 * 200, 224 * 224, 256 * 256, etc.).

Thanks for your precious time.

Best regards
